### ISSUE TITLE
Better control C++ and Java garbage collection

### ISSUE THREAD

**Issue by kwalcock** [clulab/processors#494]
As reported elsewhere, a where that is difficult to find, the master branch with glovecrash4 will run the infinite parallel test for about 2:50 hours before Java says it is out of memory.  That is with a new Processor being created each time the 10 documents are re-processed.  If the same Processor is reused, then the program ran for 20 hours without stopping.  Why?

**Comment by kwalcock** [comment]
I published a local copy of the debug version of dynet for Windows which includes a fairly good leak detector.  I ran a shortened version of the infinite parallel test, because the leak detector slows things down.  After 45 minutes the program exited and did one last garbage collection.  No leaks were detected other that the one added on purpose to make sure it is working.  It appears that all the memory is accounted for and will be collected at some time,  That time may just be too late.

**Comment by kwalcock** [comment]
Next I modified all the Scala classes in DyNet to have a close() method which calls delete on the internal Java object which in turn calls the C++ method to free memory.  In other words, manual memory management is being used in Scala.  Now only to CluProcessor.annotate() I have added code to close the embeddings.lookupParameters and embeddings.collection.  So far, the resulting program has been running 11:30 hours and the operating system does not note any excessive memory use.  Not every Scala object is being closed, but doing these few seem to help.
